pyabc/epsilon/temperature.py

```diff
@@ -0,0 +1,502 @@
+
+        self.initial_temperature = initial_temperature
+
+        self.max_nr_populations = None
+        self.temperatures = {}
+
+    def initialize(self,
+                   t: int,
+                   get_weighted_distances: Callable[[], pd.DataFrame],
+                   max_nr_populations: int,
+                   acceptor_config: dict):
+        self.max_nr_populations = max_nr_populations
+
+        # set initial temperature
+        self._update(t, get_weighted_distances, 1.0, acceptor_config)
+
+    def update(self,
+               t: int,
+               weighted_distances: pd.DataFrame,
+               acceptance_rate: float,
+               acceptor_config: dict):
+        self._update(t, lambda: weighted_distances, acceptance_rate,
+                     acceptor_config)
+
+    def _update(self,
+                t: int,
+                get_weighted_distances: Callable[[], pd.DataFrame],
+                acceptance_rate: float,
+                acceptor_config):
+        """
+        Compute the temperature for time `t`.
+        """
+        # scheme arguments
+        kwargs = dict(
+            t=t,
+            get_weighted_distances=get_weighted_distances,
+            max_nr_populations=self.max_nr_populations,
+            pdf_norm=acceptor_config['pdf_norm'],
+            kernel_scale=acceptor_config['kernel_scale'],
+            prev_temperature=self.temperatures.get(t-1, None),
+            acceptance_rate=acceptance_rate,
+        )
+
+        if t >= self.max_nr_populations - 1:
+            # t is last time
+            temperature = 1.0
+        elif not self.temperatures and self.initial_temperature is not None:
+            if callable(self.initial_temperature):
+                # execute scheme
+                temperature = self.initial_temperature(**kwargs)
+            else:
+                # should be float value
+                temperature = self.initial_temperature
+        else:
+            # evalute schemes
+            temps = []
+            for scheme in self.schemes:
+                temp = scheme(**kwargs)
+                temps.append(temp)
+            logger.debug(f"Proposed temperatures: {temps}.")
+
+            # compute next temperature based on proposals and fallback
+            # should not be higher than before
+            fallback = self.temperatures[t - 1] \
+                if t-1 in self.temperatures else np.inf
+            proposed_value = self.aggregate_fun(temps)
+            # also a value lower than 1.0 does not make sense
+            temperature = max(min(proposed_value, fallback), 1.0)
+
+        # record found value
+        self.temperatures[t] = temperature
+
+    def __call__(self,
+                 t: int) -> float:
+        return self.temperatures[t]
+
+
+class TemperatureScheme:
+    """
+    A TemperatureScheme suggests the next temperature value. It is used as
+    one of potentially multiple schemes employed in the Temperature class.
+    This class is abstract.
```

Comment:
what does this mean?

Reply:
The class being abstract? Just that this class does not implement functionality itself, it it just a base class from which all schemes are derived.

Reply:
thanks :)